openstack/networking/v2/extensions/fwaas/firewalls/requests.go

```diff
@@ -0,0 +1,169 @@
+	AdminStateUp bool   `q:"admin_state_up"`
+	Shared       bool   `q:"shared"`
+	PolicyID     string `q:"firewall_policy_id"`
+	ID           string `q:"id"`
+	Limit        int    `q:"limit"`
+	Marker       string `q:"marker"`
+	SortKey      string `q:"sort_key"`
+	SortDir      string `q:"sort_dir"`
+}
+
+// List returns a Pager which allows you to iterate over a collection of
+// firewalls. It accepts a ListOpts struct, which allows you to filter
+// and sort the returned collection for greater efficiency.
+//
+// Default policy settings return only those firewalls that are owned by the
+// tenant who submits the request, unless an admin user submits the request.
+func List(c *gophercloud.ServiceClient, opts ListOpts) pagination.Pager {
+	q, err := gophercloud.BuildQueryString(&opts)
+	if err != nil {
+		return pagination.Pager{Err: err}
+	}
+	u := rootURL(c) + q.String()
+	return pagination.NewPager(c, u, func(r pagination.PageResult) pagination.Page {
+		return FirewallPage{pagination.LinkedPageBase{PageResult: r}}
+	})
+}
+
+// CreateOpts contains all the values needed to create a new firewall.
+type CreateOpts struct {
+	// Only required if the caller has an admin role and wants to create a firewall
+	// for another tenant.
+	TenantID     string
+	Name         string
+	Description  string
+	AdminStateUp *bool
+	Shared       *bool
+	PolicyID     string
+}
+
+// ToPolicyCreateMap casts a CreateOpts struct to a map.
+func (opts CreateOpts) ToPolicyCreateMap() map[string]interface{} {
+	p := make(map[string]interface{})
+
+	if opts.TenantID != "" {
+		p["tenant_id"] = opts.TenantID
+	}
+	if opts.Name != "" {
+		p["name"] = opts.Name
+	}
+	if opts.Description != "" {
+		p["description"] = opts.Description
+	}
+	if opts.Shared != nil {
+		p["shared"] = *opts.Shared
+	}
+	if opts.AdminStateUp != nil {
+		p["admin_state_up"] = *opts.AdminStateUp
+	}
+	if opts.PolicyID != "" {
+		p["firewall_policy_id"] = opts.PolicyID
+	}
+
+	return p
+}
+
+// Create accepts a CreateOpts struct and uses the values to create a new firewall
+func Create(c *gophercloud.ServiceClient, opts CreateOpts) CreateResult {
+
+	type request struct {
+		Firewall map[string]interface{} `json:"firewall"`
+	}
+
+	reqBody := request{Firewall: opts.ToPolicyCreateMap()}
+
+	var res CreateResult
+	_, res.Err = perigee.Request("POST", rootURL(c), perigee.Options{
+		MoreHeaders: c.AuthenticatedHeaders(),
+		ReqBody:     &reqBody,
+		Results:     &res.Body,
+		OkCodes:     []int{201},
+	})
+	return res
+}
+
+// Get retrieves a particular firewall based on its unique ID.
+func Get(c *gophercloud.ServiceClient, id string) GetResult {
+	var res GetResult
+	_, res.Err = perigee.Request("GET", resourceURL(c, id), perigee.Options{
+		MoreHeaders: c.AuthenticatedHeaders(),
+		Results:     &res.Body,
+		OkCodes:     []int{200},
+	})
+	return res
+}
+
+// UpdateOpts contains the values used when updating a firewall.
+type UpdateOpts struct {
+	// Name of the firewall.
+	Name         *string
```

Comment:
Actually, this a bit more tricky because with a type `string` we are not able to determine if the field must be updated or not. It will be clear with a concrete example.

Let's say i create a firewall like this

```go
fw, err := Create(fake.ServiceClient(), CreateOpts{
	Name:         "fw",
	Description:  "OpenStack firewall",
	PolicyID:     "19ab8c87-4a32-4e6a-a74e-b77fffb89a0c",
})
```

Then, i want to update this firewall to erase the value of the `Name` field (which is legal from the API point of view). If the type is `string` the value will not be updated because it is empty. Using a `*string` we don't update the field it the value is `nil`but we update it if it is empty.

Updating the name or the description with an empty string is probably not an essential use case but if you consider the field `SourceIpAddress` in `Rule` (among others) it is much more important because updating a rule to remove the value of this field make sense and it is a common use case that must be possible.